### PR TITLE
Added custom jvp for np.linalg.slogdet

### PR DESCRIPTION
**Pull request by dpfau**
For faster gradient computation for determinants, added closed-form expressing for Jacobian-vector product. Still needs a test.

### CODE REVIEW

File: jax/numpy/linalg.py
Comment (shoyer) [inline jax/numpy/linalg.py]: will this re-use the LU decomposition from the forward pass of `slogdet`?

Comment (dpfau) [reply jax/numpy/linalg.py]: No - it would just run `solve` directly on the primal and tangent. In this case it might actually be faster to not do a custom gradient, though in general, I do wonder if it's possible to define a jvp that closes over variables in the scope of the primal computation, like you can do with `@custom_gradient` in tensorflow?

Comment (shoyer) [reply jax/numpy/linalg.py]: JAX has `custom_gradient`, too. It might be worth giving that a try.

Comment (dpfau) [reply jax/numpy/linalg.py]: True. I thought that was just syntactic sugar though, and this was the recommended way to overload derivatives.

Comment (mattjj) [reply jax/numpy/linalg.py]: I think using `custom_transforms` + `defjvp` is indeed the better way for this case, since `custom_gradient` only sets up a reverse-mode rule whereas here we're getting both a forward- and reverse-mode rule.

As for reusing the factorization, we could make that work, though this current PR is already a big improvement, and under a `jit` XLA will likely CSE things to reuse the factorization anyway. To make it save work without a `jit`, we can make a follow-up PR.

**Comment by mattjj** [comment]
I think the tests are failing due to an import issue (probably just needs a `from jax import custom_transforms`. Can you take a look?

File: jax/numpy/linalg.py
Comment (shoyer) [reply jax/numpy/linalg.py]: Right, we could use `defjvp_all` to reuse the factorization while still defining a JVP rule.

Agreed about saving that for a follow-up PR, though...

**Comment by mattjj** [comment]
@dpfau unfortunately we can't merge until the Travis CI tests pass, but it seems they're failing due to `custom_transforms` not being imported, or something. Can you take a look at that?

**Comment by dpfau** [comment]
@mattjj Sure, I still need to add tests anyway, so it's not ready to merge quite yet.

**Comment by dpfau** [comment]
I see you've added @jit to slogdet since I started this fork. Does the order of @jit and @custom_transforms matter?

**Comment by dpfau** [comment]
Should be ready to merge now, if all tests pass.

**Comment by dpfau** [comment]
What the hell...tests are passing just fine locally.

**Comment by dpfau** [comment]
Merge away!

**Review by mattjj — APPROVED** [review]
Woohoo, awesome! Thanks for this!

**Pull request merged**
